RIOT-OS/RIOT
# periph/gpio: support for extension API

### Contribution description
This is an implementation of the intercept method outlined by my proposal in [#9582](https://github.com/RIOT-OS/RIOT/issues/9582#issuecomment-410094702) (parts 1-4) for the periph/gpio interface. In short, it allows other drivers (such as GPIO expanders) to be addressed using gpio_t so that all existent code can easily gain support through those drivers. For non-implementation detail and discussion, please see #9582

At @kYc0o 's suggestion, I am breaking the implementation of #9582 / #9690 into several pieces to make it easier to review. This PR only contains the API intercept code.

### Testing procedure
A test case is provided for detecting collisions between use of ```GPIO_PIN``` and ```GPIO_EXT_PIN```. Otherwise the existing GPIO tests may be used to confirm CPU based pins work with the API change. This PR only contains API definitions for the redirection functions, so implementation of them will be tested in a follow-up PR.

### Issues/PRs references
Partially replaces and closes #9190 
Implements parts 1-4 of my proposal in #9582 
Part of work tracked by #9690
